File: 07_app_development/MyProject/libtsparser/src/main/java/com/tosmart/xiexuming/libtsparser/program/ProgramManager.java
package com.tosmart.xiexuming.libtsparser.program;

import com.tosmart.xiexuming.libtsparser.descriptor.Descriptor;
import com.tosmart.xiexuming.libtsparser.descriptor.DescriptorExtension;
import com.tosmart.xiexuming.libtsparser.descriptor.DescriptorManager;
import com.tosmart.xiexuming.libtsparser.descriptor.ServiceDescriptor;
import com.tosmart.xiexuming.libtsparser.table.ProgramAssociationTable;
import com.tosmart.xiexuming.libtsparser.table.ProgramMapTable;
import com.tosmart.xiexuming.libtsparser.table.Service;
import com.tosmart.xiexuming.libtsparser.table.ServiceDescriptionTable;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/**
 * @author xiexuming
 */
public class ProgramManager {
    private static volatile ProgramManager instance;

    private ProgramManager() {
    }

    public static ProgramManager getInstance() {
        if (instance == null) {
            synchronized (ProgramManager.class) {
                instance = new ProgramManager();
            }
        }
        return instance;
    }

    public List<Program> getProgramList(ProgramAssociationTable programAssociationTable,
                                        ServiceDescriptionTable serviceDescriptionTable,
                                        HashMap<Integer, ProgramMapTable> programMapTableMap) {
        if (programAssociationTable == null && serviceDescriptionTable == null) {
            return null;
        }
        ArrayList<Program> programs = programAssociationTable != null ? programAssociationTable.getPrograms() : null;
        ArrayList<Service> services = serviceDescriptionTable.getService();
        if (programs == null) {
            return null;
        }
        for (Program program : programs) {
            int programNumber = program.getProgramNumber();
            for (Service service : services) {
                if (programNumber != service.getServiceId()) {
                    continue;
                }
                ArrayList<Descriptor> descriptorList = service.getDescriptor();
                String serviceName = getServiceName(descriptorList);
                program.setProgramName(serviceName);
                break;
            }
            int programMapPid = program.getProgramMapPid();
            if (programMapTableMap.containsKey(programMapPid)) {
                ProgramMapTable programMapTable = programMapTableMap.get(programMapPid);
                program.setProgramMapTable(programMapTable);
            }
        }
        System.out.println(programs);
        return programs;
    }

    private String getServiceName(ArrayList<Descriptor> descriptorList) {
        DescriptorExtension descriptorExtension = new DescriptorManager();
        for (Descriptor descriptor : descriptorList) {
            ServiceDescriptor serviceDescriptor = descriptorExtension.getServiceDescriptor(descriptor);
            if (descriptor == null) {
                continue;
            }
            return serviceDescriptor.getServiceName();
        }
        return "";
    }
}

File: newpaserapp/libtosmartparse/src/main/java/com/example/libtosmartparse/program/ProgramManager.kt
import com.example.libtosmartparse.descriptor.DescriptorExtension
import com.example.libtosmartparse.descriptor.ServiceDescriptor
import com.example.libtosmartparse.table.ProgramAssociationTable
import com.example.libtosmartparse.table.Service
import com.example.libtosmartparse.table.ServiceDescriptionTable
import com.example.libtosmartparse.descriptor.Descriptor
import com.example.libtosmartparse.descriptor.DescriptorManager

/**
 * @author xxm
 * 2022/7/17
 **/
class ProgramManager private constructor() {
    fun getProgramList(
        programAssociationTable: ProgramAssociationTable?,
        serviceDescriptionTable: ServiceDescriptionTable?,
        programMapTableMap: HashMap<Int, ProgramMapTable>?
    ): List<Program>? {
        if (programAssociationTable == null && serviceDescriptionTable == null) {
            return null
        }

        val programs: ArrayList<Program>? =
            programAssociationTable?.getPrograms()
        val services: ArrayList<Service>? = serviceDescriptionTable?.getService()
        if (programs == null) {
            return null
        }
        for (program in programs) {
            val programNumber: Int = program.getProgramNumber()
            if (services != null) {
                for (service in services) {
                    if (programNumber != service.getServiceId()) {
                        continue
                    }
                    val descriptorList: ArrayList<Descriptor> = service.getDescriptor()!!
                    val serviceName = getServiceName(descriptorList)
                    program.setProgramName(serviceName)
                    break
                }
            }
            val programMapPid: Int = program.getProgramMapPid()
            if (programMapTableMap!!.containsKey(programMapPid)) {
                val programMapTable: ProgramMapTable? = programMapTableMap[programMapPid]
                program.setProgramMapTable(programMapTable)
            }
        }
        println(programs)
        return programs
    }

    private fun getServiceName(descriptorList: ArrayList<Descriptor>): String? {
        val descriptorExtension: DescriptorExtension = DescriptorManager()
        for (descriptor in descriptorList) {
            val serviceDescriptor: ServiceDescriptor =
                descriptorExtension.getServiceDescriptor(descriptor)!!
            return serviceDescriptor.getServiceName()
        }
        return "~"
    }

    companion object {
        @Volatile
        var instance: ProgramManager? = null
            get() {
                if (field == null) {
                    synchronized(this) { field = ProgramManager() }
                }
                return field
            }
            private set
    }
}
